File: Golang/Data Structures/Linked List/doubly-linked-list.go
type Any interface {
	int | int8 | int16 | int32 | int64 | uint | uint8 | uint16 | uint32 | uint64 | float32 | float64 | string
}

type Node[any Any] struct {
	Val  any
	Prev *Node[any]
	Next *Node[any]
}

type DoublyLinkedList[any Any] struct {
	Head *Node[any]
	Tail *Node[any]
	Size int
}

func (this *DoublyLinkedList[any]) Constructor() DoublyLinkedList[any] {
	doublyLinkedList := &DoublyLinkedList[any]{nil, nil, 0}

	return *doublyLinkedList
}

func (this *DoublyLinkedList[any]) getSize() int {
	return this.Size
}

func (this *DoublyLinkedList[any]) append(val any) {
	node := &Node[any]{val, nil, nil}

	if this.Head == nil {
		this.Head = node
		this.Tail = node
	} else {
		node.Prev = this.Tail
		this.Tail.Next = node
		this.Tail = this.Tail.Next
	}

	this.Size++
}

func (this *DoublyLinkedList[any]) prepend(val any) {
	node := &Node[any]{val, nil, nil}

	if this.Head == nil {
		this.Head = node
		this.Tail = node
	} else {
		this.Head.Prev = node
		node.Next = this.Head
		this.Head = node
	}

	this.Size++
}

func (this *DoublyLinkedList[any]) insert(val any, index int) {
	if index < 0 || index > this.Size {
		return
	}

	node := &Node[any]{val, nil, nil}

	if index == 0 {
		this.prepend(val)
	} else if index == this.Size {
		this.append(val)
	} else {
		current := this.Head
		i := 0

		for i < index-1 {
			current = current.Next
			i++
		}
		current.Next.Prev = node
		node.Next = current.Next
		node.Prev = current
		current.Next = node

		this.Size++
	}
}

func (this *DoublyLinkedList[any]) Print() {
	if this.Size == 0 {
		fmt.Println("[]")
	}

	current := this.Head

	for current != nil {
		fmt.Print(current.Val)
		if current.Next != nil {
			fmt.Print(" ")
		}
		current = current.Next
	}
	fmt.Println()
}

func (this *DoublyLinkedList[any]) PopRight() any {

	var errorValue int = -1

	if this.Size == 0 {
		return any(rune(errorValue))
	}

	temp := this.Head.Val

	if this.Size == 1 {
		this.Head = nil
		this.Tail = nil
		this.Size = 0
	} else {
		temp = this.Tail.Val
		this.Tail = this.Tail.Prev
		this.Tail.Next = nil
		this.Size--
	}

	return temp
}

func (this *DoublyLinkedList[any]) PopLeft() any {

	var errorValue int = -1

	if this.Size == 0 {
		return any(rune(errorValue))
	}

	temp := this.Head.Val

	if this.Size == 1 {
		this.Head = nil
		this.Tail = nil
		this.Size = 0
	} else {
		this.Head = this.Head.Next
		this.Size--
	}

	return temp
}

func (this *DoublyLinkedList[any]) Remove(index int) {

	if this.Size == 0 {
		fmt.Println("List is Empty!")
		return
	}

	if index < 0 || index > this.Size {
		fmt.Println("Invalid Index")
		return
	}

	if index == 0 {
		this.PopLeft()
	} else if index == this.getSize()-1 {
		this.PopRight()
	} else {

		current := this.Head

		currentIndex := 0

		for currentIndex < index-1 {
			current = current.Next
			currentIndex++
		}
		current.Next = current.Next.Next
		this.Size--
	}
}

func (this *DoublyLinkedList[any]) Clear() {
	this.Head = nil
	this.Tail = nil
	this.Size = 0
}

File: Java/Data Structures/Linked Lists/DoublyLinkedList.java
package com.believer.LinkedList;

class DNode <T> {
    protected T value;
    protected DNode<T> prev;
    protected DNode<T> next;

    public DNode(T value) {
        this.value = value;
        this.prev = null;
        this.next = null;
    }
}

public class DoublyLinkedList <T> {

    private DNode<T> head;
    private DNode<T> tail;
    private int size;

    public DoublyLinkedList() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public void print() {

        DNode<T> current = this.head;

        while (current != null) {
            if (current.next == null) {
                System.out.print(current.value);
            } else {
                System.out.print(current.value + " -> ");
            }
            current = current.next;
        }
    }

    public void prepend(T value) {

        DNode<T> node = new DNode<>(value);

        if (this.head == null) {
            this.head = node;
            this.tail = node;
        } else {
            node.next = this.head;
            this.head.prev = node;
            this.head = node;
        }

        this.size++;
    }

    public void append(T value) {

        DNode<T> node = new DNode<>(value);

        if (this.head == null) {
            this.head= node;
            this.tail = node;
        } else {
            node.prev = this.tail;
            this.tail.next = node;
            this.tail = this.tail.next;
        }

        this.size++;
    }

    public T popLeft() throws Exception {

        if (this.head == null) {
            throw new Exception("Cannot pop from empty list");
        }

        T popValue = this.head.value;

        if (this.head.next == null) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.next;
        }

        this.size--;

        return popValue;
    }

    public T popRight() throws Exception {


        if (this.head == null) {
            throw new Exception("Cannot pop from empty list");
        }
        T popValue = this.head.value;

        if (this.head.next == null) {
            this.head = null;
            this.tail = null;
        } else {
            this.tail = this.tail.prev;
            this.tail.next = null;
        }

        this.size--;

        return popValue;
    }

    public void remove(int index) throws Exception {

        if (this.head == null) {
            throw new Exception("Cannot remove from empty list");
        }

        if (index < 0 || index > this.size-1) {
            throw new Exception("Invalid index");
        }

        if (index == 0) {
            this.popLeft();
        } else if (index == this.size-1) {
            this.popRight();
        } else {

            DNode<T> current = this.head;
            int tempIndex = 0;

            while (tempIndex < index-1) {
                current = current.next;
                tempIndex++;
            }
            current.next = current.next.next;
            current.next.prev = null;

            this.size--;
        }
    }

    public void remove(T value) throws Exception {

        if (this.head == null) {
            throw new Exception("Cannot remove from empty list");
        }

        if (this.head.value == value) {
            this.popLeft();
        } else if (this.tail.value == value) {
            this.popRight();
        } else {

            DNode<T> current = this.head;

            while (current.next.value != value) {
                current = current.next;
            }

            current.next = current.next.next;
            current.next.prev = null;

            this.size--;
        }
    }

    public void clear() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

}
